// @ismatjon/angular-data-table/esm5/components/column/column.component.js
import * as tslib_1 from "tslib";
import { Directive, Input, ContentChild } from '@angular/core';
var DataTableColumn = /** @class */ (function () {
    function DataTableColumn() {
        this.sortable = false;
        this.resizable = false;
        this.visible = true;
        this.styleClassObject = {}; // for [ngClass]
    }
    DataTableColumn.prototype.getCellColor = function (row, index) {
        if (this.cellColors !== undefined) {
            return this.cellColors(row.item, row, this, index);
        }
    };
    DataTableColumn.prototype.ngOnInit = function () {
        this._initCellClass();
    };
    DataTableColumn.prototype._initCellClass = function () {
        var _a;
        if (!this.styleClass && this.property) {
            if (/^[a-zA-Z0-9_]+$/.test(this.property)) {
                this.styleClass = 'column-' + this.property;
            }
            else {
                this.styleClass =
                    'column-' + this.property.replace(/[^a-zA-Z0-9_]/g, '');
            }
        }
        if (this.styleClass != null) {
            this.styleClassObject = (_a = {},
                _a[this.styleClass] = true,
                _a);
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], DataTableColumn.prototype, "header", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], DataTableColumn.prototype, "sortable", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], DataTableColumn.prototype, "resizable", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], DataTableColumn.prototype, "property", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], DataTableColumn.prototype, "styleClass", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Function)
    ], DataTableColumn.prototype, "cellColors", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], DataTableColumn.prototype, "width", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], DataTableColumn.prototype, "visible", void 0);
    tslib_1.__decorate([
        ContentChild('dataTableCell'),
        tslib_1.__metadata("design:type", Object)
    ], DataTableColumn.prototype, "cellTemplate", void 0);
    tslib_1.__decorate([
        ContentChild('dataTableHeader'),
        tslib_1.__metadata("design:type", Object)
    ], DataTableColumn.prototype, "headerTemplate", void 0);
    DataTableColumn = tslib_1.__decorate([
        Directive({
            selector: 'data-table-column'
        })
    ], DataTableColumn);
    return DataTableColumn;
}());
export { DataTableColumn };
