// @ismatjon/angular-data-table/esm5/index.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { DataTable } from './components/table/table.component';
import { DataTableColumn } from './components/column/column.component';
import { DataTableRow } from './components/row/row.component';
import { DataTablePagination } from './components/pagination/pagination.component';
import { DataTableHeader } from './components/header/header.component';
import { PixelConverter } from './utils/px';
import { Hide } from './utils/hide';
import { MinPipe } from './utils/min';
import { defaultTranslations } from './types/default-translations.type';
export * from './tools/data-table-resource';
export { DataTable, DataTableColumn, DataTableRow, DataTablePagination, DataTableHeader, defaultTranslations };
export var DATA_TABLE_DIRECTIVES = [DataTable, DataTableColumn];
var DataTableModule = /** @class */ (function () {
    function DataTableModule() {
    }
    DataTableModule = tslib_1.__decorate([
        NgModule({
            imports: [CommonModule, FormsModule],
            declarations: [
                DataTable,
                DataTableColumn,
                DataTableRow,
                DataTablePagination,
                DataTableHeader,
                PixelConverter,
                Hide,
                MinPipe
            ],
            exports: [DataTable, DataTableColumn]
        })
    ], DataTableModule);
    return DataTableModule;
}());
export { DataTableModule };
